codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketServerHandshaker.java

```diff
@@ -286,18 +293,43 @@ public final ChannelFuture handshake(final Channel channel, HttpRequest req,
                 return promise;
             }
         }
-        // Add aggregator and ensure we feed the HttpRequest so it is aggregated. A limit o 8192 should be more then
-        // enough for the websockets handshake payload.
-        //
-        // TODO: Make handshake work without HttpObjectAggregator at all.
-        String aggregatorName = "httpAggregator";
-        p.addAfter(ctx.name(), aggregatorName, new HttpObjectAggregator(8192));
-        p.addAfter(aggregatorName, "handshaker", new SimpleChannelInboundHandler<FullHttpRequest>() {
+
+        String aggregatorCtx = ctx.name();
+        if (HttpUtil.isContentLengthSet(req) || HttpUtil.isTransferEncodingChunked(req) ||
+            version == WebSocketVersion.V00) {
+            // Add aggregator and ensure we feed the HttpRequest so it is aggregated. A limit o 8192 should be more then
+            // enough for the websockets handshake payload.
+            aggregatorCtx = "httpAggregator";
+            p.addAfter(ctx.name(), aggregatorCtx, new HttpObjectAggregator(8192));
+        }
+
+        p.addAfter(aggregatorCtx, "handshaker", new SimpleChannelInboundHandler<HttpObject>() {
+
+            private FullHttpRequest fullHttpRequest;
+
             @Override
-            protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest msg) throws Exception {
-                // Remove ourself and do the actual handshake
-                ctx.pipeline().remove(this);
-                handshake(channel, msg, responseHeaders, promise);
+            protected void channelRead0(final ChannelHandlerContext ctx, HttpObject msg) throws Exception {
+                if (msg instanceof FullHttpRequest) {
+                    fullHttpRequest = (FullHttpRequest) msg;
+                }
+
+                if (msg instanceof LastHttpContent) {
+                    assert fullHttpRequest != null;
+                    // Remove ourself and do the actual handshake
+                    ctx.pipeline().remove(this);
+                    handshake(channel, fullHttpRequest, responseHeaders, promise);
+                    fullHttpRequest = null;
+                    return;
+                }
+
+                if (msg instanceof HttpRequest) {
+                    HttpRequest httpRequest = (HttpRequest) msg;
+                    fullHttpRequest = new DefaultFullHttpRequest(httpRequest.protocolVersion(), httpRequest.method(),
+                        httpRequest.uri(), Unpooled.EMPTY_BUFFER, httpRequest.headers(), EmptyHttpHeaders.INSTANCE);
+                    if (httpRequest.decoderResult().isFailure()) {
+                        fullHttpRequest.setDecoderResult(httpRequest.decoderResult());
+                    }
+                }
             }
```

Comment:
```suggestion
            protected void channelRead0(final ChannelHandlerContext ctx, HttpObject msg) throws Exception {
                if (msg instanceof FullHttpRequest) {
                    fullHttpRequest = (FullHttpRequest) msg;
                }

                if (msg instanceof LastHttpContent) {
                    assert fullHttpRequest != null;
                    // Remove ourself and do the actual handshake
                    ctx.pipeline().remove(this);
                    handshake(channel, fullHttpRequest.retain(), responseHeaders, promise);
                    return;
                }

                if (msg instanceof HttpRequest) {
                    HttpRequest httpRequest = (HttpRequest) msg;
                    fullHttpRequest = new DefaultFullHttpRequest(httpRequest.protocolVersion(), httpRequest.method(),
                        httpRequest.uri(), Unpooled.EMPTY_BUFFER, httpRequest.headers(), EmptyHttpHeaders.INSTANCE);
                    if (httpRequest.decoderResult().isFailure()) {
                        fullHttpRequest.setDecoderResult(httpRequest.decoderResult());
                    }
                }
            }
            
            @Override
            public void handlerRemoved(ChannelHandlerContext ctx) {
                if (fullHttpRequest != null) {
                    fullHttpRequest.release();
                    fullHttpRequest = null;
                }
            }
```

Reply:
In this case `handlerRemoved` will be invoked before we done actual handshake.  We create FullHttpRequest with EMPTY_BUFFER if handler will be remove before we received `LastHttpContent` it not be released but for empty buffer is OK or i missed something. For the rest of cases `SimpleHttpHandler` release all requests automatically.

Reply:
But if you think it's important, I'll try to rewrite the code to use `ChannelInboundHandlerAdapter` and handle the case if a channel is removed before we receive `LastHttpContent`.

Reply:
I think we should still release

Reply:
Done